src/ape/api/transactions.py

```diff
@@ -231,10 +233,52 @@ def decode_logs(self, abi: Union[EventABI, "ContractEvent"]) -> Iterator[Contrac
         Returns:
             Iterator[:class:`~ape.types.ContractLog`]
         """
-        if not isinstance(abi, EventABI):
-            abi = abi.abi
+        if abi:
+            if not isinstance(abi, EventABI):
+                abi = abi.abi
 
-        yield from self.provider.network.ecosystem.decode_logs(abi, self.logs)
+            yield from self.provider.network.ecosystem.decode_logs(abi, self.logs)
+        else:
+            # If ABI is not provided, decode all events
+            logs_map: Dict[str, Dict[str, Tuple[EventABI, List[Dict]]]] = {}
+            for log in self.logs:
+                address = log["address"]
+                contract_type = self.chain_manager.contracts.get(address)
+
+                if not contract_type:
+                    logger.warning(f"Failed to locate contract at '{address}'.")
+                    continue
+
+                log_topics = log.get("topics", [])
+                if not log_topics:
+                    raise ValueError("Missing 'topics' in log data")
+
+                selector = log["topics"][0]
+                if selector not in contract_type.events:
+                    # Likely a library log
```

Comment:
@fubuloubu I did this to address your feedback
So now instead of looping over `self.logs` twice, it will only attempt to decode library logs if it does not locate the selector in the contact type (the heuristic).

The default impl of `def decode_library_log()` returns `None`, so it's not a breaking change.
Otherwise, ecosystems can override it to attempt to process lib logs.